source/common/upstream/eds.cc

```diff
@@ -99,47 +99,18 @@ void EdsClusterImpl::onConfigUpdate(const ResourceVector& resources) {
 
     updateHosts(current_hosts_copy, createHealthyHostList(*current_hosts_copy), per_locality,
                 createHealthyHostLists(*per_locality), hosts_added, hosts_removed);
-
-    if (initialize_callback_ && health_checker_ && pending_health_checks_ == 0) {
-      pending_health_checks_ = hosts().size();
-      ASSERT(pending_health_checks_ > 0);
-
-      // Every time a host changes HC state we cause a full healthy host recalculation which
-      // for expensive LBs (ring, subset, etc.) can be quite time consuming. During startup, this
-      // can also block worker threads by doing this repeatedly. There is no reason to do this
-      // as we will not start taking traffic until we are initialized. By blocking HC updates
-      // while initializing we can avoid this. When HC responses for all hosts have arrived and
-      // we are about to initialize, we unblock further HC updates which has the additional effect
-      // of forcing a healthy host recalculation.
-      // TODO(mattklein123): Add similar logic for the DNS clusters.
-      blockHcUpdates(true);
-      health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, bool) -> void {
-        if (pending_health_checks_ > 0 && --pending_health_checks_ == 0) {
-          blockHcUpdates(false);
-          initialize_callback_();
-          initialize_callback_ = nullptr;
-        }
-      });
-    }
+    setInitialized();
   }
 
   // If we didn't setup to initialize when our first round of health checking is complete, just
   // do it now.
-  runInitializeCallbackIfAny();
+  setInitialized();
 }
 
 void EdsClusterImpl::onConfigUpdateFailed(const EnvoyException* e) {
   UNREFERENCED_PARAMETER(e);
-  // We need to allow server startup to continue, even if we have a bad
-  // config.
-  runInitializeCallbackIfAny();
-}
-
-void EdsClusterImpl::runInitializeCallbackIfAny() {
-  if (initialize_callback_ && pending_health_checks_ == 0) {
-    initialize_callback_();
-    initialize_callback_ = nullptr;
-  }
+  // We need to allow server startup to continue, even if we have a bad config.
+  setInitialized();
```

Comment:
With our new two phased initialization is there any possibility initialization will not complete with a bad config?

Reply:
I don't believe so.